src/steady_state_stability.jl

```diff
@@ -0,0 +1,103 @@
+### Stability Analysis ###
+
+"""
+    stability(u::Vector{T}, rs::ReactionSystem, p; sparse=false, ss_jac = steady_state_jac(u, rs, p; sparse=sparse), t = Inf, non_autonomous_warn = true)
+
+Compute the stability of a steady state (Returned as a `Bool`, with `true` indicating stability).
+
+Arguments:
+    - `u`: The steady state for which we want to compute stability.
+    - `rs`: The reaction system model for which we want to compute stability.
+    - `p`: The parameter set for which we want to compute stability.
+    - `sparse=false`: If we wish to create a sparse Jacobian for the stability computation.
+    - `ss_jac = steady_state_jac(u, rs; sparse=sparse)`: It is possible to pre-compute the Jacobian used for stability computation using `steady_state_jac`. If stability is computed for many states, precomputing the Jacobian may speed up evaluation.
+    - `t = Inf`: The time point at which stability is computed. 
+    - `non_autonomous_warn = true`: If the system is non-autonomous (e.g. a rate depends on t), a warning will be given. Set this to false to remove that. Alternatively, specify a nonInf value for `t`.
+
+Example:
+```julia
+# Creates model.
+rn = @reaction_network begin
+    (p,d), 0 <--> X
+end
+p = [:p => 1.0, :d => 0.5]
+
+# Finds (the trivial) steady state, and computes stability.
+steady_state = [2.0]
+steady_state_stability(steady_state, rn, p)
+```
+
+Notes:
+    y states (each being a `Vector`) is provided as `u`, stability is computed for each state separately.
+"""
+function steady_state_stability(u::Vector{T}, rs::ReactionSystem, p; 
+                                sparse=false, ss_jac = steady_state_jac(rs; u0=u, sparse=sparse), 
+                                t = Inf, non_autonomous_warn =true) where T
+    # Warning checks.
+    !is_autonomous(rs) && non_autonomous_warn && @warn "Attempting to compute stability for a non-autonomous system. Set `non_autonomous_warn = false` to disable this warning."
+
+    # Because Jacobian currently requires u and p to be a normal vector.
+    # Can be removed once this get fixed in MTK.
+    if (u isa Vector{<:Pair}) || (u isa Dict) 
+        u_dict = Dict(symmap_to_varmap(rs, u))
+        u = [u_dict[var] for var in states(rs)]        
+    end
+    if (p isa Vector{<:Pair}) || (p isa Dict)
+        p_dict = Dict(symmap_to_varmap(rs, p))
+        p = [p_dict[var] for var in parameters(rs)]
+    end
+
+    # Computes stability (by checking that the real part of all eigenvalues are negative).
+    jac = ss_jac(u, p, Inf)
+    return maximum(real.(eigvals(jac))) < 0
```

Comment:
This needs research on what algorithms people actually use to calculate stability. I’m doubtful this is a good approach (for example, due to the difficulty in calculating all eigenvalues for large matrices accurately). I wouldn’t be in favor of adding a method that can result in misleading results due to numerical errors.

Reply:
Have you looked at any literature on methods for calculating stability numerically?

Reply:
Yeah, you mentioned that last time as well. It is a good point, I have looked around but not found anything (although simple searches typically just give university lecture intro type stuff).

I am asking around in the JuliaLab and will try to find some experts. Realistically, if it is involved, it might take a while until we can implement. In that case I'm still for merging this with appropriate warnings (as it is still a useful feature that e.g. I would have use of). But the point is good, and I haven't actually thought of it.